pkg/kube/controllers/common.go

```diff
@@ -194,17 +194,37 @@ func FromEventHandler(handler func(o Event)) cache.ResourceEventHandler {
 
 // ObjectHandler returns a handler that will act on the latest version of an object
 // This means Add/Update/Delete are all handled the same and are just used to trigger reconciling.
-func ObjectHandler(handler func(o Object)) cache.ResourceEventHandler {
+func ObjectHandler[O ComparableObject](handler func(o O)) cache.ResourceEventHandler {
 	h := func(obj any) {
-		o := ExtractObject(obj)
-		if o == nil {
+		o := Extract[O](obj)
+		if IsNil(o) {
+			return
+		}
+		handler(o)
+	}
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc: h,
+		UpdateFunc: func(oldObj, newObj any) {
+			h(newObj)
+		},
+		DeleteFunc: h,
+	}
+}
+
+// AllObjectHandler returns a handler that will act on all versions of an object
+// This means Add/Update/Delete are all handled the same and are just used to trigger reconciling.
+func AllObjectHandler[O ComparableObject](handler func(o O)) cache.ResourceEventHandler {
+	h := func(obj any) {
+		o := Extract[O](obj)
+		if IsNil(o) {
 			return
 		}
 		handler(o)
 	}
 	return cache.ResourceEventHandlerFuncs{
 		AddFunc: h,
 		UpdateFunc: func(oldObj, newObj any) {
+			h(oldObj)
```

Comment:
Why do we call twice? Can we not assume that we first `Add`ed?

Reply:
We are dealing with cross-object references. 

So
* Add Svc selecting app=foo
-time passes-
* Update Svc to app=bar selector

We need to trigger all the pods for app=foo on Update